source/load-balancers/tests/sdam-error-handling.yml

```diff
@@ -0,0 +1,192 @@
+  - database:
+      id: &multiDB multiDB
+      client: *multiClient
+      databaseName: &multiDBName multiDB
+  - collection:
+      id: &multiColl multiColl
+      database: *multiDB
+      collectionName: &multiCollName multiColl
+
+initialData:
+  - collectionName: *singleCollName
+    databaseName: *singleDBName
+    documents: []
+  - collectionName: *multiCollName
+    databaseName: *multiDBName
+    documents:
+      - _id: 1
+      - _id: 2
+      - _id: 3
+
+tests:
+  - description: only connections for a specific serverId are closed when pools are cleared
+    operations:
+      # Create two cursors to force two connections.
+      - name: createFindCursor
+        object: *multiColl
+        arguments:
+          filter: {}
+          batchSize: 2
+        saveResultAsEntity: &cursor0 cursor0
+      - name: createFindCursor
+        object: *multiColl
+        arguments:
+          filter: {}
+          batchSize: 2
+        saveResultAsEntity: &cursor1 cursor1
+      # Close both cursors to return the connections to the pool.
+      - name: close
+        object: *cursor0
+      - name: close
+        object: *cursor1
+      # Fail an operation with a state change error.
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *multiClient
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 1 }
+            data:
+              failCommands: [insert]
+              errorCode: &errorCode 11600 # InterruptedAtShutdown
+      - name: insertOne
+        object: *multiColl
+        arguments:
+          document: { x: 1 }
+        expectError:
+          errorCode: *errorCode
+      # Do another operation to ensure the relevant connection has been closed.
+      - name: insertOne
+        object: *multiColl
+        arguments:
+          document: { x: 1 }
+    expectEvents:
+      - client: *multiClient
+        events:
+          # Create cursors.
+          - connectionCreatedEvent: {}
+          - connectionReadyEvent: {}
+          - connectionCheckedOutEvent: {}
+          - connectionCreatedEvent: {}
+          - connectionReadyEvent: {}
+          - connectionCheckedOutEvent: {}
+          # Close cursors.
+          - connectionCheckedInEvent: {}
+          - connectionCheckedInEvent: {}
+          # Set failpoint.
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {}
+          # First insertOne.
+          - connectionCheckedOutEvent: {}
+          - poolClearedEvent: {}
+          - connectionCheckedInEvent: {}
+          - connectionClosedEvent:
+              reason: stale
+          # Second insertOne.
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {}
+
+  # This test uses singleClient to ensure that connection attempts are routed
+  # to the same mongos on which the failpoint is set.
+  - description: errors during the initial connection hello are ignored
```

Comment:
This test should fail when running against a 4.4 server. It needs to run against a latest server due to a failCommand bug (SERVER-49220) but cannot due to SERVER-55211.